File: backend/payment-service/src/main/java/com/skillinfinity/payment/service/PaymentService.java
package com.skillinfinity.payment.service;

import com.skillinfinity.common.dto.PageResponse;
import com.skillinfinity.payment.dto.request.PaymentConfirmationRequest;
import com.skillinfinity.payment.dto.request.PaymentFailureRequest;
import com.skillinfinity.payment.dto.request.PaymentRequest;
import com.skillinfinity.payment.dto.response.InvoiceResponse;
import com.skillinfinity.payment.dto.response.PaymentResponse;
import com.skillinfinity.payment.dto.response.ReceiptResponse;

import java.util.UUID;

public interface PaymentService {

    PaymentResponse initiatePayment(UUID userId, PaymentRequest request);

    PaymentResponse confirmPayment(UUID userId, PaymentConfirmationRequest request);

    PaymentResponse failPayment(UUID userId, PaymentFailureRequest request);

    PaymentResponse getPaymentById(UUID userId, UUID paymentId);

    PageResponse<PaymentResponse> getPaymentHistory(UUID userId, int page, int size);

    InvoiceResponse getInvoice(UUID userId, UUID invoiceId);

    ReceiptResponse getReceipt(UUID userId, UUID receiptId);
}

File: frontend/src/services/payment.service.ts
import { apiClient } from '@/api';
import { API_ENDPOINTS } from '@/constants';
import type {
  ApiResponse,
  CouponValidationRequest,
  Invoice,
  PageResponse,
  Payment,
  PaymentConfirmationRequest,
  PaymentFailureRequest,
  PaymentInitRequest,
  Receipt,
  RefundRequest,
} from '@/types';

const resolve = (template: string, params: Record<string, string>): string =>
  Object.entries(params).reduce((url, [key, value]) => url.replace(`{${key}}`, value), template);

/**
 * payment-service endpoints. Identity is resolved by the API gateway from the
 * JWT (X-User-ID header).
 */
export const paymentService = {
  initiatePayment: (payload: PaymentInitRequest) =>
    apiClient.post<ApiResponse<Payment>>(API_ENDPOINTS.PAYMENTS.INITIATE, payload),

  confirmPayment: (payload: PaymentConfirmationRequest) =>
    apiClient.post<ApiResponse<Payment>>(API_ENDPOINTS.PAYMENTS.CONFIRM, payload),

  failPayment: (payload: PaymentFailureRequest) =>
    apiClient.post<ApiResponse<Payment>>(API_ENDPOINTS.PAYMENTS.FAIL, payload),

  retryPayment: (payload: PaymentInitRequest) =>
    apiClient.post<ApiResponse<Payment>>(API_ENDPOINTS.PAYMENTS.RETRY, payload),

  requestRefund: (payload: RefundRequest) =>
    apiClient.post<ApiResponse<Payment>>(API_ENDPOINTS.PAYMENTS.REFUND, payload),

  getHistory: (page = 0, size = 20) =>
    apiClient.get<ApiResponse<PageResponse<Payment>>>(
      `${API_ENDPOINTS.PAYMENTS.HISTORY}?page=${page}&size=${size}`,
    ),

  getPayment: (paymentId: string) =>
    apiClient.get<ApiResponse<Payment>>(
      resolve(API_ENDPOINTS.PAYMENTS.ITEM, { paymentId }),
    ),

  validateCoupon: (payload: CouponValidationRequest) =>
    apiClient.post<ApiResponse<null>>(API_ENDPOINTS.PAYMENTS.COUPON, payload),

  getInvoice: (invoiceId: string) =>
    apiClient.get<ApiResponse<Invoice>>(
      resolve(API_ENDPOINTS.PAYMENTS.INVOICE, { invoiceId }),
    ),

  getReceipt: (receiptId: string) =>
    apiClient.get<ApiResponse<Receipt>>(
      resolve(API_ENDPOINTS.PAYMENTS.RECEIPT, { receiptId }),
    ),
};

export default paymentService;
